staging-service/src/main/java/com/github/cafdataprocessing/services/staging/dao/filesystem/statusreporting/BatchStatusProvider.java

```diff
@@ -0,0 +1,134 @@
+                final StagingException stagingException
+                    = new StagingException("Failed to get progress report from directory: " + inProgressBatchDir, ex);
+                progressReportExceptions.add(stagingException);
+            }
+        }
+
+        // Add in the progress reports from this service
+        inProgressReports.addAll(BatchProgressTracker.get(tenantId, batchId));
+
+        // Check if the batch has been completed
+        final boolean batchInCompletedState = batchPathProvider.getPathForBatch(tenantId, batchId).toFile().exists();
+
+        if (!batchInCompletedState) {
+            // Ignore IOExceptions that happened when retrieving the in-progress reports if the batch is completed,
+            // as they were probably due to the batch completing while the report was being collected.
+            if (!progressReportExceptions.isEmpty()) {
+                for (final StagingException ex : progressReportExceptions) {
+                    LOGGER.error("Failed to get progress report", ex);
+                }
+                throw progressReportExceptions.get(0);
+            }
+
+            if (inProgressReports.isEmpty()) {
+                throw new BatchNotFoundException(batchId.getValue());
+            }
+        }
+
+        assert batchInCompletedState || !inProgressReports.isEmpty();
+        assert batchInCompletedState || progressReportExceptions.isEmpty();
+
+        // Build and return the Response
+        final InProgress inProgress = new InProgress();
+        inProgress.setMetrics(inProgressReports.stream().map(Tracker::toInProgressMetrics).collect(Collectors.toList()));
+        final BatchStatus batchStatus = new BatchStatus();
+        batchStatus.setBatchComplete(batchInCompletedState);
+        batchStatus.setInProgress(inProgress);
+        final BatchStatusResponse response = new BatchStatusResponse();
+        response.setBatchId(batchId.getValue());
+        response.setBatchStatus(batchStatus);
+        return response;
+    }
+
+    private List<Path> getOtherServicesInProgressBatchDirectories(final TenantId tenantId, final BatchId batchId)
+    {
+        final Path tenantInProgressDir = batchPathProvider.getTenantInprogressDirectory(tenantId);
+        final String batchIdStr = batchId.getValue();
+        final List<Path> list = new ArrayList<>();
+        try (final DirectoryStream<Path> directoryStream = Files.newDirectoryStream(tenantInProgressDir)) {
+            for (final Path batch : directoryStream) {
+                final BatchInProgressDirectoryName batchDir = new BatchInProgressDirectoryName(batch.getFileName().toString());
+                if (!batchDir.isThisService() && batchDir.getBatchId().equals(batchIdStr)) {
+                    list.add(batch);
+                }
+            }
+        } catch (final IOException e) {
+            LOGGER.error("Error while traversing In-progress folder for the tenant: " + tenantId, e);
+        }
```

Comment:
> @raghavendran-gopalakrishnan Unfortunately I can't afford another large review round on this so I've applied most of my own review comments: [0f8ce49...590f331](https://github.com/CAFDataProcessing/staging-service/compare/0f8ce49de8b26e66277fce949939a17c2b5f9844...590f331c0db34ae520d5a2d8a635c62a07f8bd09). However I haven't had time to test any of the changes that I've made so this still needs to go back to you for manual testing, and possibly to look at adding some more automated tests as well. Please test everything that I've adjusted because unfortunately I haven't run it at all.

Thank you for your very extensive review and refactoring! I went through the refactored changes and manual tests seems good. I am going to check for anymore automated tests required.

Reply:
> I'm still concerned about this `IOException`. It has now been changed from `info` to `error`. The last time I asked about it the line was:
> 
> ```
> LOGGER.info("The folder might have been moved or deleted.");
> ```
> 
> ...but I want to go over it again. When does this happen? Have you ever actually seen it happening? Can you make it happen on-demand? Do you have the stack trace logged that I can look at? It is really right that we simply log the error and return an empty list?

The event of IOException happening here is not possible as it is very highly unlikely that the `in_progress` folder under the tenant folder gets deleted by any means. Perhaps that would be an error and needs to be propagated to the controller level and needs to throw an HTTP Error 500.

Reply:
If that is the situation, i.e. the `in_progress` folder itself being moved or deleted, then yes that is completely unexpected and should result in a 500.